unit_test/tstParticles.hpp

```diff
@@ -0,0 +1,140 @@
+               expected_frozen + expected_local + expected_ghost );
+}
+
+template <typename ParticlesType>
+void checkParticlePositions( const ParticlesType& particles,
+                             const std::array<double, 3> box_min,
+                             const std::array<double, 3> box_max,
+                             const std::size_t start, const std::size_t end )
+{
+    using HostAoSoA = Cabana::AoSoA<Cabana::MemberTypes<double[3], double>,
+                                    Kokkos::HostSpace>;
+    HostAoSoA aosoa_host( "host_aosoa", particles.referenceOffset() );
+    auto x_host = Cabana::slice<0>( aosoa_host );
+    auto x = particles.sliceReferencePosition();
+    Cabana::deep_copy( x_host, x );
+
+    // Check the particles were created in the right box.
+    for ( std::size_t p = start; p < end; ++p )
+        for ( int d = 0; d < 3; ++d )
+        {
+            EXPECT_GE( x_host( p, d ), box_min[d] );
+            EXPECT_LE( x_host( p, d ), box_max[d] );
+        }
+}
+
+//---------------------------------------------------------------------------//
+void testCreateParticles()
+{
+    using exec_space = TEST_EXECSPACE;
+
+    std::array<double, 3> box_min = { -1.0, -1.0, -1.0 };
+    std::array<double, 3> box_max = { 1.0, 1.0, 1.0 };
+    std::array<int, 3> num_cells = { 10, 10, 10 };
+
+    // Frozen or all particles first.
+    CabanaPD::Particles<TEST_MEMSPACE, CabanaPD::PMB,
+                        CabanaPD::TemperatureIndependent>
+        particles( exec_space(), box_min, box_max, num_cells, 0 );
+
+    // Check expected values for each block of particles.
+    std::size_t expected_local = num_cells[0] * num_cells[1] * num_cells[2];
+    std::size_t expected_frozen = 0;
+
+    checkNumParticles( particles, expected_frozen, expected_local );
+    checkParticlePositions( particles, box_min, box_max,
+                            particles.frozenOffset(), particles.localOffset() );
+}
+
+void testCreateFrozenParticles()
+{
+    using exec_space = TEST_EXECSPACE;
+
+    std::array<double, 3> box_min = { -1.0, -1.0, -1.0 };
+    std::array<double, 3> box_max = { 1.0, 1.0, 1.0 };
+    std::array<int, 3> num_cells = { 10, 10, 10 };
+
+    // Frozen in bottom half.
+    auto init_bottom = KOKKOS_LAMBDA( const int, const double x[3] )
+    {
+        if ( x[2] < 0.0 )
+            return true;
+        return false;
+    };
+    CabanaPD::Particles<TEST_MEMSPACE, CabanaPD::PMB,
+                        CabanaPD::TemperatureIndependent>
+        particles( exec_space(), box_min, box_max, num_cells, 0, init_bottom, 0,
+                   true );
+
+    // Unfrozen in top half.
+    auto init_top = KOKKOS_LAMBDA( const int, const double x[3] )
+    {
+        if ( x[2] > 0.0 )
+            return true;
+        return false;
+    };
+    // Create more, starting from the current number of frozen points.
+    particles.createParticles( exec_space{}, init_top, particles.numFrozen() );
```

Comment:
I still prefer `frozenOffset` as the most consistent name, since you may want to compare it with `localOffset` in e.g. asserts. But I won't object to this choice.